### PR TITLE
refactor: Merge similar functions `ilike_scalar` and `nilike_scalar`

### PR DESCRIPTION
**Pull request by askoa**
# Which issue does this PR close?

Part of #3296

### CODE REVIEW

**Review by tustvold — APPROVED** [review]
Might be worth just double checking the benchmarks for LLVM derp, but I'd be very surprised if this has any impact.

Thank you 😀

**Comment by askoa** [comment]
> Might be worth just double checking the benchmarks for LLVM derp, but I'd be very surprised if this has any impact.

Do benchmarks run in PR checks? Or should I run it locally?

**Comment by tustvold** [comment]
Locally, I'm also happy to run them tomorrow morning

**Review by viirya — APPROVED** [review]

**Comment by askoa** [comment]
> Locally, I'm also happy to run them tomorrow morning

That would be great!

**Comment by tustvold** [comment]
```
$ git checkout master
$ RUSTFLAGS="-C target-cpu=native" cargo bench --features="test_utils,dyn_cmp_dict" --bench comparison_kernels -- --save-baseline master ".*ilike"
...
$ git checkout like-duplication
$ RUSTFLAGS="-C target-cpu=native" cargo bench --features="test_utils,dyn_cmp_dict" --bench comparison_kernels -- --baseline master ".*ilike"
```

Resulted in

```
ilike_utf8 scalar equals
                        time:   [2.3886 ms 2.3895 ms 2.3904 ms]
                        change: [+6.2307% +6.2869% +6.3448%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

ilike_utf8 scalar contains
                        time:   [4.3580 ms 4.3595 ms 4.3610 ms]
                        change: [+5.0782% +5.1287% +5.1773%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild

ilike_utf8 scalar ends with
                        time:   [2.3695 ms 2.3705 ms 2.3715 ms]
                        change: [+0.9025% +0.9541% +1.0075%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) low mild
  2 (2.00%) high mild

ilike_utf8 scalar starts with
                        time:   [2.4795 ms 2.4811 ms 2.4827 ms]
                        change: [+5.6493% +5.7240% +5.8024%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

Benchmarking ilike_utf8 scalar complex: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 9.2s, enable flat sampling, or reduce sample count to 50.
ilike_utf8 scalar complex
                        time:   [1.8226 ms 1.8233 ms 1.8240 ms]
                        change: [-4.7478% -4.5718% -4.3951%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  4 (4.00%) low mild
  5 (5.00%) high mild
  5 (5.00%) high severe

nilike_utf8 scalar equals
                        time:   [2.4453 ms 2.4466 ms 2.4479 ms]
                        change: [+1.8160% +1.8795% +1.9439%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

nilike_utf8 scalar contains
                        time:   [4.4117 ms 4.4137 ms 4.4159 ms]
                        change: [+3.6819% +3.7352% +3.7913%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe

nilike_utf8 scalar ends with
                        time:   [2.4475 ms 2.4486 ms 2.4498 ms]
                        change: [+2.9289% +2.9881% +3.0444%] (p = 0.00 < 0.05)
                        Performance has regressed.

nilike_utf8 scalar starts with
                        time:   [2.4008 ms 2.4014 ms 2.4021 ms]
                        change: [+0.7831% +0.8124% +0.8463%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

Benchmarking nilike_utf8 scalar complex: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 9.1s, enable flat sampling, or reduce sample count to 50.
nilike_utf8 scalar complex
                        time:   [1.8052 ms 1.8061 ms 1.8071 ms]
                        change: [-6.1191% -5.9827% -5.7985%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  2 (2.00%) high severe
```

So this does represent a performance regression, I will investigate why

**Comment by tustvold** [comment]
Switching to use `BooleanArray::from_unary` made the regression go away, as well as simplifying the code :tada:

File: arrow-string/src/like.rs [hunk 1]
```diff
@@ -589,62 +588,39 @@ fn ilike_scalar_op<'a, F: Fn(bool) -> bool, L: ArrayAccessor<Item = &'a str>>(
...
-                    bit_util::set_bit(bool_slice, i);
-                }
-            }
-        }
+        Ok(BooleanArray::from_unary(left, |item| {
+            op(item.to_uppercase().ends_with(ends_str))
+        }))
     } else if right.starts_with('%')
         && right.ends_with('%')
+        && !right.ends_with("\\%")
```
Comment (tustvold) [inline arrow-string/src/like.rs:616]: This appears to have been an omission from https://github.com/apache/arrow-rs/pull/2743

**Pull request merged**